src/concurrency/async/state-machine.md

```diff
@@ -0,0 +1,97 @@
+---
+minutes: 7
+---
+
+# State Machine
+
+Rust transforms an async function or block to a hidden type that implements
+`Future`, using a state machine to track the function's progress. The details of
+this transform are complex, but it helps to have a schematic understanding of
+what is happening.
+
+```rust,editable,compile_fail
+use futures::executor::block_on;
+use std::future::Future;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+async fn send(s: &str) {
+    println!("{s}");
+}
+
+/*
+async fn count_to(count: i32) {
+    for i in 1..=count {
+        send("tick").await;
+    }
+}
+*/
+
+fn count_to(count: i32) -> CountToFuture {
+    CountToFuture { state: CountToState::Init, count, i: 0 }
+}
+
+struct CountToFuture {
+    state: CountToState,
+    count: i32,
+    i: i32,
+}
+
+enum CountToState {
+    Init,
+    Sending(Pin<Box<dyn Future<Output = ()>>>),
+}
+
+impl std::future::Future for CountToFuture {
+    type Output = ();
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        loop {
+            match &mut self.state {
+                CountToState::Init => {
+                    self.i = 1;
+                    self.state = CountToState::Sending(Box::pin(send("tick")));
+                }
+                CountToState::Sending(send_future) => {
+                    match send_future.as_mut().poll(cx) {
+                        Poll::Pending => return Poll::Pending,
+                        Poll::Ready(_) => {
+                            self.i += 1;
+                            if self.i > self.count {
+                                return Poll::Ready(());
+                            } else {
+                                self.state =
+                                    CountToState::Sending(Box::pin(send("tick")));
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+fn main() {
+    block_on(count_to(5));
+}
+```
+
+<details>
+
+While this code will run, it is simplified from what the real state machine
+would do. The important things to notice here are:
+
+- Calling an async function does nothing but construct a value, ready to start
+  on the first call to `poll`.
+- All local variables are stored in the function's future struct, including an
+  enum to identify where execution is currently suspended. The real generated
+  state machine would not initialize `i` to 0.
```

Comment:
I think it makes more sense for explanation to first trace the control flow so we can enumerate states and get in our heads something like "we'll have three possible execution states for this async fn, so it makes sense that its future type would be (morally) an enum with three variants". Then we can look at variable liveness at each of the awaits to determine the payload of each variant. This lets us get to "futures store live locals" without having to introduce the notion of liveness explicitly like a compilers course would--these are just the variables whose values we'll need to run the rest of the function.

I think it's better for an example of the state machine transform to use an async fn that doesn't have async in a loop, so that it's easy for readers to enumerate the entire set of states in their head.

So I might suggest an example more like this:
```rust
async fn send(s: String) -> usize {
    println!("{}", s);
    s.len()
}

async fn example(x: i32) -> usize {
    let double_x = x*2;
    let mut count = send(format!("x = {x}")).await;
    count += send(format!("double_x = {double_x}")).await;
    count
}
```

This gives us three states:
- an initial state holding the argument `x`
- a state holding `double_x` until we return from our first `await`
- a state holding `count` until we return from our second `await`

This is, I think, less confusing than a state we return to across iterations of our loop on `1..=count`.

Reply:
Implicit in this suggestion is to use enum variants to represent the liveness of variables. That's great, but does require a bit more mucking about with pins than we want to present at this point. In fact, it uses `pin_project`, which we don't even talk about in the Pin section. Should I maybe revert this to a flat struct with a `fut` field, just to avoid this?

Reply:
I don't think it's so important to make this example have perfect fidelity or be compilable/runnable--pseudocode to capture the transformation would be fine, and would allow us to side-step `Pin` questions. The sub-future, if we really want to represent it, could just be a `std::future::Ready` or isomorphic. The most significant thing is not the state of the child future but the fact that at await points we capture live state.